cvmfs/kvstore.cc

```diff
@@ -0,0 +1,233 @@
+             "miss%s on GetSize",
+             id.ToString().c_str());
+    return -ENOENT;
+  }
+}
+
+int64_t MemoryKvStore::GetRefcount(const shash::Any &id) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_getrefcount);
+  if (entries_.Lookup(id, &mem)) {
+    // LogCvmfs(kLogKvStore, kLogDebug, "%s has refcount %u",
+    //          id.ToString().c_str(), mem.refcount);
+    return mem.refcount;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on GetRefcount",
+             id.ToString().c_str());
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::IncRef(const shash::Any &id) {
+  perf::Inc(counters_.n_incref);
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (entries_.Lookup(id, &mem)) {
+    assert(mem.refcount < UINT_MAX);
+    ++mem.refcount;
+    entries_.Insert(id, mem);
+    LogCvmfs(kLogKvStore, kLogDebug, "increased refcount of %s to %u",
+             id.ToString().c_str(), mem.refcount);
+    return true;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on IncRef",
+             id.ToString().c_str());
+    return false;
+  }
+}
+
+bool MemoryKvStore::Unref(const shash::Any &id) {
+  perf::Inc(counters_.n_unref);
+  WriteLockGuard guard(rwlock_);
+  MemoryBuffer mem;
+  if (entries_.Lookup(id, &mem)) {
+    assert(mem.refcount > 0);
+    --mem.refcount;
+    entries_.Insert(id, mem);
+    LogCvmfs(kLogKvStore, kLogDebug, "decreased refcount of %s to %u",
+             id.ToString().c_str(), mem.refcount);
+    return true;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on Unref",
+             id.ToString().c_str());
+    return false;
+  }
+}
+
+int64_t MemoryKvStore::Read(
+  const shash::Any &id,
+  void *buf,
+  size_t size,
+  size_t offset
+) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_read);
+  ReadLockGuard guard(rwlock_);
+  if (entries_.Lookup(id, &mem)) {
+    if (offset > mem.size) {
+      LogCvmfs(kLogKvStore, kLogDebug, "out of bounds read (%u>%u) on %s",
+               offset, mem.size, id.ToString().c_str());
+      return 0;
+    }
+    uint64_t copy_size = min(mem.size - offset, size);
+    // LogCvmfs(kLogKvStore, kLogDebug, "copy %u B from offset %u of %s",
+    //          copy_size, offset, id.ToString().c_str());
+    memcpy(buf, static_cast<char *>(mem.address) + offset, copy_size);
+    perf::Xadd(counters_.sz_read, copy_size);
+    return copy_size;
+  } else {
+    LogCvmfs(kLogKvStore, kLogDebug, "miss %s on Read", id.ToString().c_str());
+    return -ENOENT;
+  }
+}
+
+bool MemoryKvStore::Commit(
+  const shash::Any &id,
+  const kvstore::MemoryBuffer &buf
+) {
+  MemoryBuffer mem;
+  perf::Inc(counters_.n_commit);
+  WriteLockGuard guard(rwlock_);
+  LogCvmfs(kLogKvStore, kLogDebug, "commit %s", id.ToString().c_str());
+  if (entries_.Lookup(id, &mem)) {
+    LogCvmfs(kLogKvStore, kLogDebug, "commit overwrites existing entry");
+    used_bytes_ -= mem.size;
+    perf::Xadd(counters_.sz_deleted, mem.size);
+    counters_.sz_size->Set(used_bytes_);
```

Comment:
The awkward layout here is so that it's safe to commit or open from transaction. I added some comments to clarify